### ISSUE TITLE
Towards baseline probabilistic predictions

### ISSUE THREAD

**Issue by michaelchughes** [tufts-ml/covid19-forecasting#28]
We have two basic (and very related) prediction tasks:

### Task 1: Admission

Given admission counts from the past W timesteps, predict next admission count

That is, we want to learn a distribution
$$
p( a_t | a_{t-W:t-1} )
$$
We could also condition on external features $x$ (e.g. the forecasted state-wide admissions from some trustworthy model).

### Task 2: Census

Given census counts from the past W timesteps, predict next census count

$$
p( c_t | c_{t-W:t-1} )
$$
Of course, we could also condition on other features (e.g. also on emissions, or on census counts from another stage). But let's focus on this for now.

# Learning goals

If $t$ is today's date, we want to make a probabilistic forecast for 1 week ahead: $a_{t+1}, a_{t+2}, \ldots a_{t+7}$

# Model strategy

For either task, let's try to get each one of these working before we move on to the next step:

1) Consider a Bayesian linear regression model with Normal likelihood

Use Stan or PyMC3 (or roll your own, if needed)

2) Allow for *missing* data in the context

That is, we want to be able to deal with the fact that some windows $W$ will not have all observed values. 

3) Consider a Bayesian generalized linear model with an integer-sample-space likelihood

We are predicting *counts*. We should use something like Poisson, Negative Binomial, etc.
Using Gaussian is naive and wastes probability mass on non-integer values.

4) Look at fancier models (e.g. Gaussian process, linear model with learned basis function, etc.)

# Problems we'll need to solve along the way

Hyperparameter selection:
* How to select likelihood hypers (variance, window size)?
* How to select prior hypers?

**Comment by michaelchughes** [comment]
What would a solution look like?

Imagine we have a csv file on disk like this one:
https://github.com/tufts-ml/covid19-forecasting/blob/master/nyt_data/counties/Middlesex.csv

Would like interface to be something like this:

```
python run_simple_forecast.py --input_csv_file /path/to/data.csv --target_col_name cases --model_file model.json --output_csv_file_pattern /path/to/output-{sample}.csv --n_samples 5000 --day_forecasts_start DATE --n_days_ahead 14
```

With the output being a bunch of CSV files containing forecasted counts from date DATE to n_days_ahead. 

We can then use our built in summarize_forecasts.py file to summarize the forecasts

**Comment by ally-lee** [comment]
I made the changes we talked about, so the model is now regressing directly on the previous W counts, where the prior coefficient for the most recent timestep has a mean of 1 and all others have a mean of 0. I also tried using difference features, but that seemed to give very similar performance to using the counts themselves.

I switched from Normal to Poisson likelihood which improved performance. I'm getting a heldout log likelihood (per day) of around -5 at best (with Normal likelihood, it was around -7). What log likelihood should I be aiming for?

I'm making these plots for future predictions with error bars of +- 1.96 * sqrt(mu). Is this the best way to calculate the confidence interval? I know you mentioned that the error bars should increase as we go farther into the future. They do increase, simply because the predicted counts increase over time, but not because of increased uncertainty.

Can you explain what --model_file and --n_samples mean in your specification for the interface?

![Figure_1](https://user-images.githubusercontent.com/30029312/83811796-5708da00-a66f-11ea-94e1-39353b46a540.png)
![Figure_3](https://user-images.githubusercontent.com/30029312/83811802-583a0700-a66f-11ea-9f10-8228f23bbf40.png)

**Comment by michaelchughes** [comment]
Surprised you see such big gains moving from Normal to Poisson. How are you setting the variance of the Normal?

Probably we do want a likelihood that (like the Poisson) only puts mass on integers. However, we also want a likelihood that allows controlling its mean and variance with separate parameters (Poisson uses one number to define both).
One easy option that fits both criteria is a negative binomial likelihood.
https://stats.idre.ucla.edu/stata/seminars/regression-models-with-count-data/

Error bars should come from sampling from the distribution. That is, your model can give you a distribution: p(y_t | history), for each future timestep y_t. If you sample a y_t count value 1000 times from it, you can take the 2.5th percentile and the 97.5th percentile of those 1000 samples as your interval of uncertainty (since 95% of the probability mass is contained between those values). This kind of estimate is useful because it is NOT symmetric (e.g. if you have a lopsided distribution, one bound may be much further from the mean than another)

## Parameters

Model file should be a JSON file that specifies any hyperparameters needed (e.g. if you are fixing the window size, or fixing the variance of a Normal likelihood)

n_samples should be the number of samples we draw from the posterior predictive p(y_t | history).

**Comment by ally-lee** [comment]
Is this the right way to predict future counts?
p(y_t | y[t-W : t-1]) = NegativeBinomial(mu = bias + inner_prod(beta, y[t-W : t-1]), alpha = alpha)

1. Use find_MAP to learn a set of parameters for a negative binomial distribution (above).
2. Take 1000 samples from this distribution. Use the 2.5th and 97.5th percentile as the bounds of an interval estimate. Should the point estimate be the median of the samples or the mean of the distribution itself?
3. Add point estimate for y_t to the set of "previous" counts and repeat step 2 for y_{t+1}.

I tried playing around with the prior distributions and got the performance with the Normal to be comparable to that of the Poisson. The prior for sigma is now HalfNormal(sd = 10). Before I had sd=1 which was too small.

**Comment by michaelchughes** [comment]
1) Definitely good idea to tune any variance hyperparameter (sigma for Normal, alpha for NegativeBinomial), or the prior for that hyperparameter, on the validation set (pick one that maximizes the likelihood of validation set). Note that Poisson doesn't have a free hyperparameter for variance.

2) When you make predictions for y_t+1, use Monte Carlo integration to average across samples of previous predictions at y_t

**Comment by ally-lee** [comment]
I'm having trouble understanding how to average across samples from the previous timestep and use that for future sampling. When I take the average of the 1000 samples, this value is almost equivalent to the mean of the distribution, so I'm tempted to just plug in that point estimate to compute the mean for future predictions. I'm also wondering why it's necessary to take samples when we can analytically compute the 2.5th and 97.5th percentiles of the distribution. Could you recommend something for me to read to better understand this task?

**Comment by michaelchughes** [comment]
Here’s a quick explanation. Make sense?

Let’s let “Y” denote random variable, and “y” as observed variable

You want to draw samples from the *joint* distribution over T future time steps given W past time steps.

p(Y_t+1, Y_t+2, … Y_t+T | y_t, … y_t-W )


To do this, use the product rule

1) sample from y_t+1 ~ p( Y_t+1 | y_t, … y_t-W )
2) sample from y_t+2 ~ p( Y_t+2 | y_t+1 …. y_{t-W+1} )   [uses sample from 1]
3) sample from y_t+3 ~ p( Y_t+3 | y_t+2, y_t+1, … ) [uses sample from 1 and 2]
...

 

On Jun 10, 2020, at 3:53 PM, ally-lee <notifications@github.com> wrote:


I'm having trouble understanding how to average across samples from the previous timestep and use that for future sampling. When I take the average of the 1000 samples, this value is almost equivalent to the mean of the distribution, so I'm tempted to just plug in that point estimate to compute the mean for future predictions. I'm also wondering why it's necessary to take samples when we can analytically compute the 2.5th and 97.5th percentiles of the distribution. Could you recommend something for me to read to better understand this task?

**Comment by ally-lee** [comment]
This makes sense! Thank you!

**Comment by ally-lee** [comment]
I just created a new branch for myself and uploaded what I've been working on in a folder called baseline-predictions. It's still a work in progress so the full functionality isn't there, but I wanted to get something up before the meeting tomorrow. Here's a quick summary:

1. holdout_last20.py: Takes in CSV file of counts. Searches a set of hyperparameters and uses the last 20% of examples as validation set (i.e., most recent timesteps). Produces a JSON file with the best model parameters. Produces a plot of heldout log likelihood vs window size.

    - Right now the script just searches a set of window sizes, but I can add other parameter spaces to search. I've played around with the prior hyperparameters (coefficients for linear regression, alpha parameter for negative binomial), so I'm just defaulting to the best ones I've found for the Middlesex data set.

2. run_simple_forecast.py: Takes in CSV file of counts, JSON file with model specification, and a few other optional arguments. Trains model using given data and hyperparameters. Takes samples of predictive distribution for next n_days_ahead and writes forecasts to CSV files. Produces plot of mean of samples for each day with error bars depicting 2.5th percentile and 97.5th percentile.

**Comment by ally-lee** [comment]
I cleaned up and documented my code and uploaded the new versions. If you have time to take a look and test it, that would be great!

How to run:
1. `python grid_search.py Middlesex.csv`
2. `python run_simple_forecast.py Middlesex.csv`

Notes:

- I switched from using find_MAP to taking samples from the posterior and then using the mean as my parameter estimate. find_MAP was giving wonky results after I switched from HalfNormal to TruncatedNormal prior for alpha. This takes longer but I think it’s safer.

- The grid search takes a while - around 3 minutes to train 8 models. Feel free to change the hyperparameter search spaces (lines 65-68 of grid_search.py).

- I’ve uploaded a file called model.json that has the “best” model I’ve found for Middlesex.csv, so that can be plugged directly into run_simple_forecast.py. The program defaults to reading/writing a file named 'model.json'.

- run_simple_forecast.py writes each sample to a CSV file. I have a subdirectory called 'samples', so the program defaults to 'samples/output-*.csv' as the output CSV file pattern. You can change this using the -o option. These files can be used as input to summarize_forecasts.py.

- It defaults to taking 1000 samples, but the Negative Binomial sampler is really slow, so you can change the number of samples using the -n option. It will output its progress every 10 samples.

- You can change the start date of forecasts using the -s option, as long as the date is at least (W+1) days after the earliest data point and at most 1 day after the last data point.

- You can change the number of days (after the start date) to make forecasts for using the -d option.

Next steps for me:
1. Work on handling missing data.
2. Start testing on our dataset.

**Comment by ally-lee** [comment]
Questions:
1. I’m working on figuring out how to handle missing data. It’s pretty easy to do if just y values are missing, but this problem is tricky since the y values are also used as predictor variables, so we end up having missing values throughout both the x and y datasets. Do you have thoughts on how to work around this? There is a way to sample unknown x values, but that doesn’t seem ideal since the missing x and y values should be exactly the same values.
2. Can you confirm which dataset I should be working with? Will the format be exactly like StandardizedCounts.xlsx?

**Comment by michaelchughes** [comment]
FYI I'm trying out your code now!

The quick feedback I have is that we should do a *monte carlo* estimate of the "score", rather than a plug-in estimate based on a single sample of the posterior mean.


That is, we should do:

$$
p( y_t+1 | y_{t-W:t}, \lambda) = \sum_{s=1}^S p( y_t+1 | y_{t-W:t}, \lambda, \theta^s)
$$

where

$$
\theta^s \sim p( \theta | y_{train} )
$$

**Comment by michaelchughes** [comment]
RE missing data

I think instead of manually breaking up the time series into "y_train, x_train" before we define the model, it may make more sense to just provide a single long time series y_train.

For example, see here:
https://discourse.pymc.io/t/vectorized-autoregressive-model/1449

Then, should be possible to just provide a "nan" value for missing entries, and pymc3 will sample those

**Comment by michaelchughes** [comment]
@ally-lee no rush, but whenever you get a chance do keep me in loop when you ping pymc3 help forums to try to get an answer about the missing data issues...

**Comment by ally-lee** [comment]
I created a notebook that summarizes the approaches I've tried for handling missing data. I also included a draft of what I'll post on the PyMC3 help forum.
https://github.com/tufts-ml/covid19-forecasting/blob/al-baseline-predictions/baseline-predictions/missing_data.ipynb

**Comment by ally-lee** [comment]
I posted my question, and I'll keep you updated on what they say!
https://discourse.pymc.io/t/handling-missing-data-in-linear-regression-on-timeseries-data/5364

**Comment by michaelchughes** [comment]
Great! Thanks for filing that issue. Looks good to me

**Comment by michaelchughes** [comment]
Thanks for this hard work ally! closing...
